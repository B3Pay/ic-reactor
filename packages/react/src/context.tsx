import React, {
  createContext,
  useState,
  useEffect,
  PropsWithChildren,
  useMemo,
  useCallback,
} from "react"
import { ActorSubclass, HttpAgent } from "@dfinity/agent"
import { createReActor } from "./index"
import { IDL } from "@dfinity/candid"
import { getDidJsFromMetadata, getDidJsFromTmpHack } from "@ic-reactor/store"
import { ReActorOptions } from "@ic-reactor/store"

export type ReActor<A> = ReturnType<typeof createReActor<A>>

export interface ReActorContextType<A = ActorSubclass<any>>
  extends ReActor<A> {}

const ReActorContext = createContext<ReActorContextType | null>(null)

export const useReActor = <
  A = ActorSubclass<any>,
>(): ReActorContextType<A> => {
  const context = React.useContext(ReActorContext)

  if (!context) {
    throw new Error("useReActor must be used within a ReActorProvider")
  }

  return context
}

interface ReActorProviderProps
  extends PropsWithChildren,
    Omit<ReActorOptions, "idlFactory"> {
  idlFactory?: IDL.InterfaceFactory
  loadingComponent?: React.ReactNode
}

export const ReActorProvider: React.FC<ReActorProviderProps> = ({
  children,
  loadingComponent = <div>Loading...</div>,
  ...config
}) => {
  const [didJs, setDidJS] = useState<{ idlFactory: IDL.InterfaceFactory }>()

  const fetchDidJs = useCallback(async () => {
    const { canisterId } = config
    if (!canisterId) {
      throw new Error("canisterId is required")
    }

    const agent = new HttpAgent({ ...config })

    getDidJsFromMetadata(agent, canisterId).then(async (idlFactory) => {
      if (!idlFactory) {
        try {
          idlFactory = await getDidJsFromTmpHack(agent, canisterId)
        } catch (err) {
          if (/no query method/.test(err as any)) {
            console.warn(err)
            idlFactory = undefined
          } else {
            throw err
          }
        }

        if (!idlFactory) {
          console.warn("No query method found for canister", canisterId)
        }
      }
      setDidJS(idlFactory)
    })

    return didJs
  }, [config.canisterId])

  useEffect(() => {
    const { idlFactory } = config

    if (idlFactory) {
      setDidJS({ idlFactory })
      return
    }

    fetchDidJs()
  }, [fetchDidJs, config.idlFactory])

  const reActorHooks = useMemo(() => {
    if (!didJs) {
      return null
    }
    try {
      return createReActor<any>({
        ...config,
        idlFactory: didJs.idlFactory,
      })
    } catch (err) {
      console.error(err)
      return null
    }
  }, [config, didJs])

  return (
    <ReActorContext.Provider value={reActorHooks}>
      {reActorHooks ? children : loadingComponent}
    </ReActorContext.Provider>
  )
}
