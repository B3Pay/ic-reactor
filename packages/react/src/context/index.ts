export * from "./actor"
export * from "./agent"
