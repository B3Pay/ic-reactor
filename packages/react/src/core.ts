export {
  createReactorStore,
  createActorManager,
  createAgentManager,
  createCandidAdapter,
} from "@ic-reactor/core"
