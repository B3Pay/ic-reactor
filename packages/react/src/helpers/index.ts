export * from "./getAgent"
export * from "./getActor"
export * from "./getAuth"
export * from "./extractActor"
export * from "./extractAgent"
