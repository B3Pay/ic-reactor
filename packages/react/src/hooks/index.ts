export * from "./useActor"
export * from "./useCandid"
export * from "./agent"
export * from "./actor"
export * from "./auth"
