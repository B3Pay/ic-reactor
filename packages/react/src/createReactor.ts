import { createReactorStore } from "@ic-reactor/core"
import { agentHooks, authHooks, actorHooks } from "./helpers"

import type {
  BaseActor,
  CreateReactorParameters,
  CreateReactorReturnType,
} from "./types"

/**
 * Initializes and configures the reactor environment for interacting with the Internet Computer (IC) blockchain within a React application.
 * It encapsulates the creation of actors, authentication, and agent management, offering a streamlined interface for blockchain interactions.
 *
 * @param config Configuration config for the reactor, including:
 *  - `withProcessEnv` (optional): Specifies whether to use process environment variables to determine if the environment is local or development. Defaults to false.
 *  - `isLocalEnv` (optional): Indicates if the current environment is local or development, influencing the agent and actor behavior. Useful for testing or development.
 *  - `port` (optional): Port number for the local or development environment.
 *  Extends `CreateReactorStoreParameters` which includes HTTP agent config, actor manager config (excluding `agentManager`), and an optional custom agent manager.
 *
 * @returns An object containing various hooks and utilities:
 *  - {@link getAgent} - Returns the current agent instance.
 *  - {@link getVisitFunction} - Returns the visit function for the actor.
 *  - {@link useQueryCall} - A hook for querying actor methods.
 *  - {@link useUpdateCall} - A hook for updating actor methods.
 *  - {@link useAuth} - A hook for managing authentication and user principal.
 *  - {@link useActor} - A hook for managing actors and their methods.
 *  - {@link useActorManager} - A hook for managing actor manager and its methods.
 *  - {@link useAgentManager} - A hook for managing agent manager and its methods.
 *  - {@link initialize} - A function to initialize the actor manager if not initialized.
 *  - {@link useActorState} - A hook for managing actor state.
 *  - {@link useAgent} - A hook for managing agent and its methods.
 *  - {@link useAuthState} - A hook for managing authentication state.
 *  - {@link useAgentState} - A hook for managing agent state.
 *  - {@link useUserPrincipal} - A hook for managing user principal.
 *  - {@link useVisitMethod} - A hook for visiting actor methods.
 *
 * @example
 * ```typescript
 * import { createReactor } from "@ic-reactor/react";
 * import type { CreateReactorParameters } from "@ic-reactor/react/dist/types";
 * import { canisterId, idlFactory, yourActor } from "./declaration/yourActor"
 *
 * const config: CreateReactorParameters = {
 *   canisterId,
 *   idlFactory,
 *   host: "https://localhost:8000", // IC network host         |
 *   isLocalEnv: true, // Set true for local network            | one of these
 *   withProcessEnv: true, // Use process.env to determine host |
 *   port: 8000, // Port number for local network               |
 * };
 *
 * export type YourActor = typeof yourActor;
 *
 * export const { useAuth, useQueryCall, useUpdateCall } = createReactor<YourActor>(config);
 * // Now you can use the returned hooks in your React components
 * ```
 */
export const createReactor = <A = BaseActor>(
  config: CreateReactorParameters
): CreateReactorReturnType<A> => {
  const actorManager = createReactorStore<A>(config)

  const getVisitFunction = () => {
    return actorManager.visitFunction
  }

  const getAgent = () => {
    return actorManager.agentManager.getAgent()
  }

  return {
    getAgent,
    getVisitFunction,
    ...actorHooks(actorManager),
    ...authHooks(actorManager.agentManager),
    ...agentHooks(actorManager.agentManager),
  }
}
