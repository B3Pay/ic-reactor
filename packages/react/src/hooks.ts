import {
  ActorSubclass,
  CallMethod,
  ExtractReActorMethodArgs,
  ReActorActorStore,
} from "@ic-reactor/store"
import { useCallback, useEffect, useMemo, useRef, useState } from "react"
import {
  ReActorCallArgs,
  ReActorHookState,
  ReActorUseQueryArgs,
  ReActorUseUpdateArgs,
} from "./types"
import { useStore } from "zustand"

export const getCallHooks = <A extends ActorSubclass<any>>(
  callMethod: CallMethod<A>,
  actorStore: ReActorActorStore<A>
) => {
  const useMethodFields = () => {
    return useStore(actorStore, (state) => state.methodFields)
  }

  const useMethodField = (functionName: keyof A & string) => {
    const methodFields = useMethodFields()

    const field = useMemo(() => {
      return methodFields.find((f) => f.functionName === functionName)
    }, [methodFields, functionName])

    return field
  }

  const useReActorCall = <M extends keyof A>({
    onError,
    onSuccess,
    onLoading,
    args = [] as unknown as ExtractReActorMethodArgs<A[M]>,
    functionName,
  }: ReActorCallArgs<A, M>) => {
    const [state, setState] = useState<ReActorHookState<A, M>>({
      data: undefined,
      error: undefined,
      loading: false,
    })

    const call = useCallback(
      async (replaceArgs?: ExtractReActorMethodArgs<A[M]>) => {
        onLoading?.(true)
        onError?.(undefined)
        setState((prevState) => ({
          ...prevState,
          loading: true,
          error: undefined,
        }))

        try {
          const data = await callMethod(functionName, ...(replaceArgs ?? args))

          onLoading?.(false)
          onSuccess?.(data)
          setState((prevState) => ({ ...prevState, data, loading: false }))

          return data
        } catch (error) {
          console.error("Error in call:", error)
          onError?.(error as Error)
          onLoading?.(false)
          setState((prevState) => ({
            ...prevState,
            error: error as Error,
            loading: false,
          }))
        }
      },
      [args, functionName, onError, onSuccess, onLoading]
    )

    const field = useMethodField(functionName)

    return { call, field, ...state }
  }

  const useQueryCall = <M extends keyof A>({
    autoRefresh,
    refreshInterval = 5000,
    disableInitialCall,
    ...rest
  }: ReActorUseQueryArgs<A, M>) => {
    const { call, ...state } = useReActorCall(rest)

    let intervalId = useRef<NodeJS.Timeout | undefined>(undefined)

    useEffect(() => {
      // Auto-refresh logic
      if (autoRefresh) {
        intervalId.current = setInterval(() => {
          call()
        }, refreshInterval)
      }

      // Initial call logic
      if (!disableInitialCall) {
        call()
      }

      return () => {
        clearInterval(intervalId.current)
      }
    }, [disableInitialCall, autoRefresh, refreshInterval])

    return { call, ...state }
  }

  const useUpdateCall = <M extends keyof A>(
    args: ReActorUseUpdateArgs<A, M>
  ) => {
    return useReActorCall(args)
  }

  return {
    useQueryCall,
    useUpdateCall,
    useMethodField,
    useMethodFields,
  }
}
