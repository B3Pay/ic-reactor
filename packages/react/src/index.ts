import {
  createReActorStore,
  CreateReActorOptions,
  BaseActor,
} from "@ic-reactor/core"
import { getActorHooks } from "./hooks/actor"
import { getAuthHooks } from "./hooks/auth"

export { createReActor as createReActorCore } from "@ic-reactor/core"
export * from "@ic-reactor/core"

export * from "./context/agent"
export * from "./context/actor"
export * from "./hooks"

export const createReActor = <A = BaseActor>({
  isLocalEnv,
  withVisitor,
  withProcessEnv,
  ...options
}: CreateReActorOptions) => {
  isLocalEnv =
    isLocalEnv ||
    (withProcessEnv
      ? typeof process !== "undefined" &&
        (process.env.DFX_NETWORK === "local" ||
          process.env.NODE_ENV === "development")
      : false)

  const actorManager = createReActorStore<A>({
    isLocalEnv,
    withVisitor,
    ...options,
  })

  const getVisitFunction = () => {
    if (!withVisitor) {
      throw new Error(
        "Service fields not initialized. Pass `withVisitor` to initialize service fields."
      )
    }

    return actorManager.visitFunction
  }

  const getAgent = () => {
    return actorManager.agentManager.getAgent()
  }

  const actorHooks = getActorHooks<A>(actorManager)
  const authHooks = getAuthHooks(actorManager.agentManager)

  return {
    getAgent,
    getVisitFunction,
    ...actorHooks,
    ...authHooks,
  }
}
