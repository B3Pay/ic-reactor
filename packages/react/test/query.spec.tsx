import React from "react"
import renderer, { act } from "react-test-renderer"
import { createReActor } from "../src"
import { backend, idlFactory } from "./candid"

describe("createReActor", () => {
  it("should query on mount", async () => {
    const { useQueryCall } = createReActor<typeof backend>({
      canisterId: "xeka7-ryaaa-aaaal-qb57a-cai",
      idlFactory,
    })

    const TestComponent = () => {
      const { data, loading } = useQueryCall({
        functionName: "version",
      })

      return (
        <div>
          <span>
            {loading ? "Loading..." : data ? data.toString() : "Ready To call"}
          </span>
        </div>
      )
    }

    let screen = renderer.create(<TestComponent />)

    const versionStatus = () => screen.root.findAllByType("span")[0]

    expect(versionStatus().props.children).toEqual("Ready To call")
    expect(screen.toJSON()).toMatchSnapshot()

    await act(async () => {
      await new Promise((r) => setTimeout(r, 100))
    })

    expect(versionStatus().props.children).toEqual("Loading...")
    expect(screen.toJSON()).toMatchSnapshot()

    await act(async () => {
      await new Promise((r) => setTimeout(r, 1000))
    })

    expect(versionStatus().props.children).toEqual("0.2.0")
    expect(screen.toJSON()).toMatchSnapshot()
  })

  it("should query manually", async () => {
    const { useQueryCall } = createReActor<typeof backend>({
      canisterId: "xeka7-ryaaa-aaaal-qb57a-cai",
      idlFactory,
    })

    const TestComponent = () => {
      const { call, data, loading } = useQueryCall({
        functionName: "version",
        refetchOnMount: false,
      })

      return (
        <div>
          <button onClick={() => call()}>Get Version</button>
          <span>
            {loading ? "Loading..." : data ? data.toString() : "Ready To call"}
          </span>
        </div>
      )
    }

    let screen = renderer.create(<TestComponent />)

    const versionStatus = () => screen.root.findAllByType("span")[0]

    const versionButton = () => screen.root.findAllByType("button")[0]

    expect(screen.toJSON()).toMatchSnapshot()

    expect(versionStatus().props.children).toEqual("Ready To call")

    await act(() => versionButton().props.onClick())

    expect(versionStatus().props.children).toEqual("0.2.0")

    expect(screen.toJSON()).toMatchSnapshot()
  })
})
