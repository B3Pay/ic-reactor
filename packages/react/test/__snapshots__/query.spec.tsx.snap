// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createReActor should query manually 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Get Version
  </button>
  <span>
    Ready To call
  </span>
</div>
`;

exports[`createReActor should query manually 2`] = `
<div>
  <button
    onClick={[Function]}
  >
    Get Version
  </button>
  <span>
    0.2.0
  </span>
</div>
`;

exports[`createReActor should query on mount 1`] = `
<div>
  <span>
    Ready To call
  </span>
</div>
`;

exports[`createReActor should query on mount 2`] = `
<div>
  <span>
    Loading...
  </span>
</div>
`;

exports[`createReActor should query on mount 3`] = `
<div>
  <span>
    0.2.0
  </span>
</div>
`;
